lib/ace/ext/code_lense.js

```diff
@@ -0,0 +1,197 @@
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL AJAX.ORG B.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+define(function(require, exports, module) {
+"use strict";
+var LineWidgets = require("../line_widgets").LineWidgets;
+var dom = require("../lib/dom");
+
+function clearLenseElements(renderer) {
+    var textLayer = renderer.$textLayer;
+    var lenseElements = textLayer.$lenses;
+    if (lenseElements)
+        lenseElements.forEach(function(el) {el.remove(); });
+    textLayer.$lenses = null;
+}
+
+function renderWidgets(changes, renderer) {
+    var changed = changes & renderer.CHANGE_LINES 
+        || changes & renderer.CHANGE_FULL 
+        || changes & renderer.CHANGE_SCROLL
+        || changes & renderer.CHANGE_TEXT;
+    if (!changed) 
+        return;
+    
+    var session = renderer.session;
+    var lineWidgets = renderer.session.lineWidgets;
+    var textLayer = renderer.$textLayer;
+    var lenseElements = textLayer.$lenses;
+    if (!lineWidgets) {
+        if (lenseElements)
+            clearLenseElements(renderer);
+        return;
+    }
+
+    var textCells = renderer.$textLayer.$lines.cells;
+    var config = renderer.layerConfig;
+    var padding = renderer.$padding;
+
+    if (!lenseElements)
+        lenseElements = textLayer.$lenses = [];
+
+
+    var index = 0;
+    for (var i = 0; i < textCells.length; i++) {
+        var row = textCells[i].row;
+        var widget = lineWidgets[row];
+        var lenses = widget && widget.lenses;
+
+        if (!lenses || !lenses.length) continue;
+
+        var lenseContainer = lenseElements[index];
+        if (!lenseContainer) {
+            lenseContainer = lenseElements[index]
+                = dom.buildDom(["div", {class: "ace_codeLense"}], renderer.container);
+        }
+        lenseContainer.style.height = config.lineHeight + "px";
+        index++;
+
+        for (var j = 0; j < lenses.length; j++) {
+            var el = lenseContainer.childNodes[2 * j];
+            if (!el) {
+                if (j != 0) lenseContainer.appendChild(dom.createTextNode("\xa0|\xa0"));
+                el = dom.buildDom(["a"], lenseContainer);
+            }
+            el.textContent = lenses[j].title;
```

Comment:
Looks like we are safe with respect to XSS. Are there any other places where we insert user controlled content into the DOM?

Reply:
no, this is the only new place.